nemo/collections/nlp/models/nlp_model.py

```diff
@@ -410,3 +522,160 @@ def save_checkpoint(self, filepath, weights_only: bool):
+                    # Restored from .nemo, checkpoint_version will already be set
+                    pass
+                elif trainer.resume_from_checkpoint is not None:
+                    # PTL auto-resuming, need to update checkpoint name
+                    # update path based on model parallel rank
+                    filepath = trainer.resume_from_checkpoint
+                    dirname = os.path.dirname(os.path.dirname(filepath))
+                    basename = os.path.basename(filepath)
+                    filepath = f'{dirname}/mp_rank_{app_state.model_parallel_rank:02d}/{basename}'
+                    trainer.resume_from_checkpoint = filepath
+                    logging.info(f'Resuming training from checkpoint {trainer.resume_from_checkpoint}')
+                    # need to set checkpoint version for megatron-lm
+                    checkpoint_version = torch.load(trainer.resume_from_checkpoint).get('checkpoint_version', None)
+                    if checkpoint_version is not None:
+                        set_checkpoint_version(checkpoint_version)
+                    else:
+                        logging.warning('Megatron-lm checkpoint version not found. Setting checkpoint_version to 0.')
+                        set_checkpoint_version(0)
+                else:
+                    logging.info(
+                        f"Restoring from pretrained model parallel checkpoint: {self.lightning_module.bert_model._restore_path}"
+                    )
+                    self.lightning_module.bert_model.restore_weights(self.lightning_module.bert_model._restore_path)
+
+            self.lightning_module.register_megatron_checkpoint_version()
+
+        return super().start_training(trainer)
+
+    def start_testing(self, trainer: 'Trainer') -> None:
+        """ PTL Hook that is called after DPP is initialized. """
+        app_state = AppState()
+
+        if app_state.model_parallel_size is not None:
+
+            if isinstance(self.lightning_module.bert_model, MegatronBertEncoder):
+                # check megatron checkpoint version
+                checkpoint_version = get_checkpoint_version()
+                if checkpoint_version is None:
+                    raise ValueError("Unable to find megatron checkpoint version.")
+
+        return super().start_testing(trainer)
+
+    def configure_ddp(self):
+        """ Override LightningModule ddp if using model parallel.
+            Sets find_unused_parameters to True.
+        """
+
+        app_state = AppState()
+
+        if app_state.model_parallel_size is not None:
+            logging.info(f"Configuring DDP for model parallelism.")
+
+            # With model parallelism, multiple GPUs form a large "logical GPU"
+            # this means that data parallel groups span multiple GPUs
+            # and are non-trivial
+            device_ids = self.determine_ddp_device_ids()
+            self._model = DistributedDataParallel(
```

Comment:
Im afraid future versions of PTL signature for this code will change, and it will cause issues. Theres a way to inspect signatures of classes using `inspect` ive used for comparing classes and their respective dataclasses, can a similar technique be used here to ensure that signature doesnt diverge too much ?

Reply:
Another good suggestions, but can we defer to a later PR?